File: arrays_hashing/valid_anagram/cpp/valid_anagram.cpp
#include<iostream>
#include<string>

using std::cout;
using std::string;

// Time O(n)
// Space O(n)

class Solution {
public:
static bool isAnagram(string s, string t) {
    int index[256] = { 0 }; // initialize to zero

    if(s.size() != t.size()) return false;

    for(int i = 0; i < s.size(); i++) {
        index[s[i]]++;
        index[t[i]]--;
    }
    for(int i = 0; i < 256; i++) {
        if(index[i] != 0)
            return false;
    }
    return true;
}
};

int main() {
    string s = "anagram";
    string t = "nagaram";

    cout << std::boolalpha << Solution::isAnagram(s, t) << '\n';
}

File: arrays_hashing/valid_anagram/js/valid_anagram.js
const s = 'anagram';
const t = 'nagaram';

const isAnagram = (s, t) => {
  if(s.length !== t.length) return false;

  let index = Array(256).fill(0);

  for(let i in s) {
    index[s[i].charCodeAt(0)]++;
    index[t[i].charCodeAt(0)]--;
  }
  return index.filter(c => c !== 0).length === 0;
}
console.log(isAnagram(s, t));
